File: config/awesome/widget/net.lua
local awful = require("awful")
local textbox = require("widget/lib/textbox")

local function update_netstat()
    local function format_width(num)
        speed = num / 1024
        if speed > 99 then
            return string.format('%4dKb', math.floor(speed))
        else
            return string.format('%4.1fKb', speed)
        end
    end

    local interval = net_widget_clock.timeout
    local netif, text
    for line in io.lines('/proc/net/route') do
        netif = line:match('^(%w+)%s+00000000%s')
        if netif then
            break
        end
    end

    if netif then
        local down, up
        for line in io.lines('/proc/net/dev') do
            -- Match wmaster0 as well as rt0 (multiple leading spaces)
            local name, recv, send = string.match(line, "^%s*(%w+):%s+(%d+)%s+%d+%s+%d+%s+%d+%s+%d+%s+%d+%s+%d+%s+%d+%s+(%d+)")
            if name == netif then
                if netdata[name] == nil then
                    -- Default values on the first run
                    netdata[name] = {}
                    down, up = 0, 0
                else
                    down = (recv - netdata[name][1]) / interval
                    up   = (send - netdata[name][2]) / interval
                end
                netdata[name][1] = recv
                netdata[name][2] = send
                break
            end
        end
        text = ' 🔻<span color="#c2ba62">'.. format_width(down) ..'</span> 🔺<span color="#5798d9">'.. format_width(up) ..'</span>'
    else
        netdata = {} -- clear as the interface may have been reset
        text = '(No network)'
    end
    net_widget:set_markup(text .. " ")
end
netdata = {}
net_widget = textbox('(net)')
net_widget_clock = timer({ timeout = 1 })
net_widget_clock:connect_signal("timeout", update_netstat)
net_widget_clock:start()
update_netstat()

return net_widget

File: config/qtile/widget/net.py
from datetime import datetime

from libqtile.widget import base


class Net(base.InLoopPollText):

    def __init__(self):
        super().__init__("", update_interval=1)
        self.name = ''
        self.recv = 0
        self.send = 0
        self.last = datetime.now().timestamp()

    def format_net(self, num: float) -> str:
        num = num / 1024
        if num > 1024:
            return f'<span weight="bold">{int(num/1024):3} MB</span>'
        else:
            return f'{int(num):3} KB'

    def netstat(self):
        netif = ''
        for line in open('/proc/net/route').readlines():
            net, dest = line.split()[:2]
            if dest == '00000000':
                netif = net
        if not netif:
            raise Exception('not found')

        recv, send = 0, 0
        for line in open('/proc/net/dev').readlines():
            data = line.split()
            if data[0][:-1] != netif:
                continue
            recv = int(data[1])
            send = int(data[9])
        if recv == 0 or send == 0:
            raise Exception('get data failed')

        now = datetime.now().timestamp()
        down = (recv - self.recv) / (now - self.last)
        up = (send - self.send) / (now - self.last)

        self.recv = recv
        self.send = send
        self.last = now

        if self.name != netif:
            self.name = netif
            raise Exception('first init')

        result = ''
        result += f'🔻<span color="#c2ba62">{self.format_net(down)}</span>'
        result += f'🔺<span color="#5798d9">{self.format_net(up)}</span>'
        return result

    def poll(self):
        try:
            return self.netstat()
        except Exception:
            return '(no network)'
